### PR TITLE
Simplify DistributedExecutionPlanner to SplitSourceFactory

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/split/SplitSourceProvider.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.split;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.metadata.TableLayoutHandle;
+import com.facebook.presto.spi.connector.ConnectorSplitManager.SplitSchedulingStrategy;
+
+public interface SplitSourceProvider
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/split/SplitSourceProvider.java]: Do we really need this interface ? Looks like we can just pass `SplitManager` to `CloseableSplitSourceProvider` right ? :D

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 2]
```diff
@@ -316,8 +322,9 @@ private static void updateQueryOutputLocations(QueryStateMachine queryStateMachi
         Optional<int[]> bucketToPartition;
         PartitioningHandle partitioningHandle = plan.getFragment().getPartitioning();
         if (partitioningHandle.equals(SOURCE_DISTRIBUTION)) {
+            Map<PlanNodeId, SplitSource> splitSources = splitSourceFactory.createSplitSources(plan.getFragment(), session);
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: Per discussion, this is a refactor so we still make split source created eagerly when all stage execution are created. Do we want to add a TODO or comment explain ? ditto for line 345.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java [hunk 3]
```diff
@@ -469,14 +465,13 @@ private void planDistribution(PlanRoot plan)
         stateMachine.beginDistributedPlanning();
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java]: Note the method name, comment and state machine are still using the word `distributed plan`. Maybe left a TODO so we can cleanup them later?

Also, the comment `time distribution planning` looks interesting. Are we expecting distributed planning (i.e. creates all the split sources) taking some significant amount of time?

File: presto-main/src/main/java/com/facebook/presto/split/SplitSourceProvider.java (hunk 1 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/split/SplitSourceProvider.java]: I wanted to encapsulate the `CloseableSplitSourceProvider#close` method from anywhere else, but the place where the `CloseableSplitSourceProvider` is created. So it is easier to reason about the code, as it is guaranteed that the `CloseableSplitSourceProvider` won't be closed anywhere else in the code.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java (hunk 3 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java]: I'm going to remove the distributed planning time as a separate commit / PR.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 4]
```diff
@@ -316,8 +322,10 @@ private static void updateQueryOutputLocations(QueryStateMachine queryStateMachi
         Optional<int[]> bucketToPartition;
         PartitioningHandle partitioningHandle = plan.getFragment().getPartitioning();
         if (partitioningHandle.equals(SOURCE_DISTRIBUTION)) {
+            // TODO: open split sources only when scheduling
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: do you mean `defer opening split sources when stage execution starts`?

File: presto-main/src/main/java/com/facebook/presto/split/CloseableSplitSourceProvider.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    public synchronized void close()
+    {
+        if (!closed) {
+            closed = true;
+            for (SplitSource source : splitSources) {
+                try {
+                    source.close();
+                }
+                catch (Throwable t) {
+                    log.warn(t, "Error closing split source");
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/split/CloseableSplitSourceProvider.java]: I think the code before repackaged the exception and would throw.  Do we really want to swallow the exception here?

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/split/CloseableSplitSourceProvider.java]: I guess this is the same as the current implementation in `DistributedExecutionPLanner`

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/split/CloseableSplitSourceProvider.java]: I think the idea was to keep the way it was: https://github.com/prestodb/presto/pull/12582#discussion_r272238406

CC: @wenleix

**Comment by wenleix** [comment]
Historically `DistributedExecutionPlanner` is responsible for deriving certain execution-related properties, such as partitioning (see the file 7 years ago: https://github.com/prestodb/presto/blob/ff623e753bc1aadf6945e1a076f7a62174090b8b/presto-main/src/main/java/com/facebook/presto/sql/planner/DistributedExecutionPlanner.java) . As time evolves, these responsibility are moved out (e.g. partitioning reasoning are moved to `PlanFragment`), and now `DistributedExecutionPlanner` is only used to creates split sources -- which is also the only difference between `SubPlan` and `StageExecutionPlan`.

**Pull request merged**